File: src/main/java/com/epam/mentoring/kotlin/rest/DogBreedApiClient.java
package com.epam.mentoring.kotlin.rest;

import com.epam.mentoring.kotlin.model.DogBreedApiResponse;
import java.util.List;
import java.util.Map;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.reactive.function.client.WebClient;
import reactor.core.publisher.Flux;

@Component
public class DogBreedApiClient {
//    private static final String DOG_BREED_API_URL = "https://dog.ceo/api/breeds/list/all";
    private static final String DOG_BREED_API_URL = "https://dog.ceo";
//    private final RestTemplate restTemplate;
    private final WebClient webClient;

    //    public DogBreedApiClient(final RestTemplate restTemplate) {
//        this.restTemplate = restTemplate;
//    }
    public DogBreedApiClient(final WebClient.Builder webClientBuilder) {
        this.webClient = webClientBuilder.baseUrl(DOG_BREED_API_URL).build();

    }

    public Map<String, List<String>> getBreeds() throws Exception {
//        ResponseEntity<DogBreedApiResponse> response = restTemplate.getForEntity(DOG_BREED_API_URL, DogBreedApiResponse.class);

        final ResponseEntity<DogBreedApiResponse> response =
                webClient.get()
                        .uri("/api/breeds/list/all")
                        .retrieve()
                        .toEntity(DogBreedApiResponse.class)
                        .block();


        if(200 != response.getStatusCode().value()) {
            throw new Exception();
        }
        final DogBreedApiResponse body = response.getBody();
        return body.message();  //Map<String, List<String>> message
    }


}

File: src/main/java/com/epam/mentoring/kotlin/rest/DogBreedApiClient.kt
package com.epam.mentoring.kotlin.rest

import com.epam.mentoring.kotlin.model.DogBreedApiResponse
import com.epam.mentoring.kotlin.model.ImageUrlResponse
import org.springframework.stereotype.Component
import org.springframework.web.reactive.function.client.WebClient
import org.springframework.web.reactive.function.client.awaitBody

@Component
class DogBreedApiClient(webClientBuilder: WebClient.Builder) {

    private val DOG_BREED_API_URL = "https://dog.ceo"
    private val webClient: WebClient = webClientBuilder.baseUrl(DOG_BREED_API_URL).build()

    @Throws(Exception::class)
    suspend fun getBreeds(): Map<String, List<String>> {

        val response = webClient.get()
            .uri("/api/breeds/list/all")
            .retrieve()
            .toEntity(DogBreedApiResponse::class.java)
            .block()

        if (200 != response.statusCode.value()) {
            throw Exception()
        }
        val body = response.body
        return body.message             //Map<String, List<String>> message
    }

    suspend fun getImage(breed: String): ByteArray? {
        var imagesUrls: List<String>? = getImagesUrls(breed)
        var image: ByteArray? = downloadImage(imagesUrls)

        return image
    }

    private suspend fun getImagesUrls(breed: String): List<String>? {
        val response: ImageUrlResponse = webClient.get()
            .uri("/api/breed/$breed/images")
            .retrieve()
            .awaitBody<ImageUrlResponse>()

        if (response == null) {
            throw Exception()
        }
        return response.message
    }


    private suspend fun downloadImage(imagesUrls: List<String>?): ByteArray? {
        var url: String? = imagesUrls
                                    ?.filter { it.isNotBlank() }
                                    ?.first()

        var image: ByteArray? = webClient.get()
            .uri(url)
            .retrieve()
            .awaitBody<ByteArray>()

//            takeIf { it.statusCode.is2xxSuccessful }
//            ?.body

        if (image == null) {
            throw Exception()
        }

        return image
    }

}
